### PR TITLE
User accounts

### PR DESCRIPTION
**Pull request by jschr**
This PR adds user accounts to the frontend (login, register, password reset and confirm email).

- [x] Login
- [x] Register
- [x] Confirm email
- [x] Send password reset email
- [x] Change password 
- [x] Logout
- [x] SSR auth storage (cookies)

**NextJS Auth Example**
https://github.com/trandainhan/next.js-example-authentication-with-jwt

**CSRF**
I dont think we’ll need to worry about CSRF attacks because the frontend does not do any state changing requests (ie changing email), that’s all done via the API which doesn’t use cookies.

If we add routes for posting in the future (ie. to allow users to submit forms without javascript) we will need to make sure we are protected with a csrf token.

### CODE REVIEW

**Comment by KristerV** [comment]
you are fast.. Wasn't expecting this so soon. With this tempo v1 will be released pretty soon :)

Also I like the syntax you're using. Just `const { view } = this.state` is better than my usual `const view = this.state.view`. Details :D

edit: oh wait one more thing. I didn't write any of this down, but I'm super hoping we can have all of the site work without JavaScript, if the user chooses so. We've got next.js with all the fancy live updating we need, but it's also got server side rendering. So if your component is using state to the point that it doesn't work without it, then I'd appreciate it if it used less. 

For example it's fine if the JS enabled people will get their username automatically moved between login and register forms, but JS disabled people don't. But both should be able to login and register :)

**Comment by jschr** [comment]
I can work on that but I don't think any of the buttons on the site will work without javascript since we are using ajax requests instead of form post requests which return new html. 

My understanding is we can offer read-only capabilities without javascript but none of the forms will work as they all rely on click handlers set by React. That means the login / register forms won't work unless we use the form post model. 

Am I correct in that assumption?

**Comment by KristerV** [comment]
oh yeah.. how do you submit a form in react without javascript? :D And even if you do the server can't process it anyway, it's not preapared.

Okay let's just leave it for now until someone complains.

File: package.json [hunk 1]
```diff
@@ -4,29 +4,31 @@
   "description": "Helium Pay budget voting website front-end",
   "main": "index.js",
   "scripts": {
-    "dev": "next -p 5000",
     "build": "next build",
-    "start": "next start -p 5000",
+    "start": "node index.js",
```
Comment (KristerV) [inline package.json]: Why are we going against [what next.js suggests](https://github.com/zeit/next.js#stable)?

Also I didn't remember that production is [a separate process](https://github.com/zeit/next.js#production-deployment), good reminder.

File: components/Menu.js [hunk 2]
```diff
@@ -13,6 +15,23 @@ export default props => (
 					<a>Create proposal</a>
 				</Link>
 			</li>
+			{isLoggedIn ? (
+				<li>
+					<Link href="/" prefetch>
+						<div>
+							<a href="javascript:;" onClick={cookieUtils.remove}>
```
Comment (KristerV) [inline components/Menu.js:21]: Does "javascript:;" actually change anything or is it just a placeholder?

File: pages/changePassword.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import React from 'react'
+import Link from 'next/link'
+import router from 'next/router'
+import moment from 'moment'
+import Bitcore from 'bitcore-lib-dash'
```
Comment (KristerV) [inline pages/changePassword.js]: Seems you've got unused dependencies. I think we may need to move to WebStorm to have this optimization automated. But I don't think either of our laptops are powerful enough :D

File: pages/changePassword.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+		if (password !== confirmPassword) {
+			this.setState({ error: 'passwords do not match', isFetching: false })
+		} else {
+			try {
+				// uses reset token to change password
+				const client = new ApiClient(config.apiUrl, token)
+				const user = await client.resetPassword(userId, password)
+				// create and sets new user token
+				const userToken = await client.login(user.username, password)
```
Comment (KristerV) [inline pages/changePassword.js:46]: So if login fails (password doesn't match) it'll throw exception? I'll boot up the code soon too, just leaving a note for now.

File: pages/changePassword.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+												onChange={this.setFormValue('confirmPassword')}
+											/>
+										</td>
+									</tr>
+								</tbody>
+							</table>
+							<br />
+							{error && <Alert>{error}</Alert>}
+							<button type="submit" disabled={isFetching}>
+								{isFetching ? 'Updating...' : 'Update Password'}
```
Comment (KristerV) [inline pages/changePassword.js:97]: Would be awesome to have a context sensitive button component eventually. I love those things.

File: pages/submit.js [hunk 6]
```diff
@@ -4,22 +4,30 @@ import moment from 'moment'
...
+			endepoch: moment()
+				.add(2, 'months')
+				.unix()
 		}
-		this.createPrepareCommand = this.createPrepareCommand.bind(this)
-		this.createSubmitCommand = this.createSubmitCommand.bind(this)
 	}
 
-	createPrepareCommand(e) {
+	createPrepareCommand = e => {
```
Comment (KristerV) [inline pages/submit.js:29]: oh I see. didn't know one could do that :) Doesn't it still detach the context though?

File: package.json (hunk 1 above)
Comment (KristerV) [reply package.json]: I tested it and it runs. But perhaps there's some code that doesn't get updated when changed?

**Review by KristerV — COMMENTED** [review]
Ran it too, but registering responded with "Failed to fetch". Not sure what's missing.

File: package.json (hunk 1 above)
Comment (jschr) [reply package.json]: Our index.js wasn't actually running before. This is how you start next js with a custom server. I'm setting up extra middleware and we may want to add more routes in the future.

Custom server example: https://github.com/zeit/next.js/blob/master/examples/custom-server/package.json#L7

I think I'll change it to the above example though for dev / prod.

File: components/Menu.js (hunk 2 above)
Comment (jschr) [reply components/Menu.js:21]: You need some href value to get `cursor: pointer` on hover. `javascript:;` doesn't actually change anything.

File: pages/changePassword.js (hunk 3 above)
Comment (jschr) [reply pages/changePassword.js]: This should be easily fixable now that eslint is added. I thought it was the default to warn on unused dependencies but I'll take a look.

File: pages/changePassword.js (hunk 4 above)
Comment (jschr) [reply pages/changePassword.js:46]: Yeah, the client will reject the promise for any non success status code.

File: pages/changePassword.js (hunk 5 above)
Comment (jschr) [reply pages/changePassword.js:97]: Definitely! I didn't spend any time really on the UI side, just hooking the api up. I'd like to improve our form components generally and clean things up a bit.

File: pages/submit.js (hunk 6 above)
Comment (jschr) [reply pages/submit.js:29]: This is equivalent to binding in the constructor. It's using class properties + arrow functions to bind the this to the current instance of the class.

**Comment by jschr** [comment]
Is the backend started as well? You'll get "failed to fetch" if it can't hit the api which should be at localhost:3000. Also double check your config.js that apiUrl, it should be localhost. Wondering if we should be committing the config and just pull from env vars for the api url?

**Pull request merged**